datafusion/sql/src/relation/join.rs

```diff
@@ -107,7 +110,20 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
             JoinConstraint::On(sql_expr) => {
                 let join_schema = left.schema().join(right.schema())?;
                 // parse ON expression
-                let expr = self.sql_to_expr(sql_expr, &join_schema, planner_context)?;
+                let mut expr =
+                    self.sql_to_expr(sql_expr, &join_schema, planner_context)?;
+                // ON expression must be a boolean expression
+                match expr.get_type(&join_schema)? {
+                    DataType::Boolean => {}
+                    DataType::Null => {
+                        expr = Expr::Cast(Cast::new(expr.into(), DataType::Boolean))
+                    }
+                    other => {
+                        return plan_err!(
+                            "ON expression must be boolean, but got {other:?}"
```

Comment:
I think the same rule applies to other APIs (not just SQL). 

What would you think about moving the check to `LogicalPlanBuilder` 

https://github.com/apache/datafusion/blob/6e637488188c6620ecd113bf47987bd98f6d7871/datafusion/expr/src/logical_plan/builder.rs#L671-L686

That way this check would be applied to the DataFrame API as well as any custom plans that were created?

Reply:
Done. I re-implemented it in `type_coercion` to avoid creating a temporary [joined schema](https://github.com/apache/datafusion/blob/1dfac86a89750193491cf3e04917e37b92c64ffa/datafusion/sql/src/relation/join.rs#L108). Also added a test for DataFrame.